### PR TITLE
Remember the source of interactively defined functions

### PR DESCRIPTION
**Pull request by rhymeswithcycle**
- Patches the linecache module to achieve this
- inspect.getsource() works for functions defined on the console
- tracebacks show code for frames inside interactively defined functions

fixes bpython/bpython#229

### CODE REVIEW

File: bpython/history.py [hunk 1]
```diff
@@ -232,3 +233,74 @@ def append_reload_and_write(self, s, filename, encoding):
                 # Make sure that entries contains at least one element. If the
                 # file and s are empty, this can occur.
                 self.entries = ['']
+
+class BPythonLinecache(dict):
+    """Replaces the cache dict in the standard-library linecache module,
+    to also remember (in an unerasable way) bpython console input."""
+
+    def __init__(self, *args, **kwargs):
+        dict.__init__(self, *args, **kwargs)
```
Comment (sebastinas) [inline bpython/history.py]: `dict` is new style class, isn't it? Please use `super` instead.

File: bpython/history.py [hunk 2]
```diff
@@ -232,3 +233,74 @@ def append_reload_and_write(self, s, filename, encoding):
...
+linecache.cache = BPythonLinecache(linecache.cache)
+linecache.clearcache = _bpython_clear_linecache
+
+def filename_for_console_input(code_string):
+    """Remembers a string of source code, and returns
+    a fake filename to use to retrieve it later."""
+    try:
+        return linecache.cache.remember_bpython_input(code_string)
+    except AttributeError:
+        return '<input>'
```
Comment (sebastinas) [inline bpython/history.py]: Is there a reason for catching the `AttributeError`s? We've set it to our implementation after all.

File: bpython/history.py [hunk 3]
```diff
@@ -232,3 +233,74 @@ def append_reload_and_write(self, s, filename, encoding):
                 # Make sure that entries contains at least one element. If the
                 # file and s are empty, this can occur.
                 self.entries = ['']
+
+class BPythonLinecache(dict):
```
Comment (sebastinas) [inline bpython/history.py]: I think I'd prefer this in a separate file. `bpython/history.py` is for handling the history file.

**Comment by rhymeswithcycle** [comment]
Sure, moved things to a separate file.

The motivation for catching `AttributeError`s is if some other code also patches the same object -- if that happens, better to just no longer have `inspect.getsource()` work for interactively-defined functions than to raise an exception.

**Pull request merged**